modules/src/ics04_channel/handler/chan_open_try.rs

```diff
@@ -0,0 +1,479 @@
+                        old_channel_end
+                            .counterparty()
+                            .channel_id()
+                            .clone()
+                            .unwrap()
+                            .to_string(),
+                    ),
+                ))
+            }
+        }
+        // No channel id is supplied to  create a new channel end. Note: the id is assigned
+        // by the ChannelKeeper.
+        None => {
+            channel_id = None;
+
+            Ok(ChannelEnd::new(
+                State::Init,
+                *msg.channel.ordering(),
+                msg.channel.counterparty().clone(),
+                msg.channel.connection_hops().clone(),
+                msg.counterparty_version().clone(),
+            ))
+        }
+    }?;
+
+    // An IBC connection running on the local (host) chain should exist.
+
+    if msg.channel.connection_hops().len() != 1 {
+        return Err(Kind::InvalidConnectionHopsLength.into());
+    }
+
+    let connection_end = ctx.connection_state(&msg.channel().connection_hops()[0]);
+
+    let conn = connection_end
+        .ok_or_else(|| Kind::MissingConnection(msg.channel().connection_hops()[0].clone()))?;
+
+    if !conn.state_matches(&ConnectionState::Open) {
+        return Err(ConnectionNotOpen(msg.channel.connection_hops()[0].clone()).into());
+    }
+
+    let get_versions = conn.versions();
+    let version = match get_versions.as_slice() {
+        [version] => version,
+        _ => return Err(Kind::InvalidVersionLengthConnection.into()),
+    };
+
+    let channel_feature = msg.channel().ordering().as_string().to_string();
+    if !version.is_supported_feature(channel_feature) {
+        return Err(Kind::ChannelFeatureNotSuportedByConnection.into());
+    }
+
+    //Channel capabilities
+    let cap = ctx.port_capability(&msg.port_id().clone());
+    let cap_key;
+
+    match cap {
+        Some(key) => {
+            if !ctx.capability_authentification(&msg.port_id().clone(), &key) {
+                return Err(Kind::InvalidPortCapability.into());
+            } else {
+                cap_key = key;
+            }
+        }
+        None => return Err(Kind::NoPortCapability.into()),
+    }
+
+    if msg.channel().version().is_empty() {
+        return Err(Kind::InvalidVersion.into());
+    }
+
+    // Proof verification in two steps:
+    // 1. Setup: build the Channel as we expect to find it on the other party.
+
+    let _expected_counterparty = Counterparty::new(msg.port_id().clone(), None);
+
+    let counterparty = conn.counterparty();
+    let ccid = counterparty.connection_id().ok_or_else(|| {
+        Kind::UndefinedConnectionCounterparty(msg.channel().connection_hops()[0].clone())
+    })?;
+
+    let expected_connection_hops = match ctx.connection_state(ccid) {
+        Some(_c) => vec![ccid.clone()],
+        None => return Err(MissingConnectionCounterparty(ccid.clone()).into()),
+    };
```

Comment:
I don't understand this, connection with `ccid` is not stored in `ctx` as it is the counterparty connection id.
```suggestion
    let expected_connection_hops = vec![ccid.clone()];
```

Reply:
It's a verification they do in the go code: it checks that the connection received in the message (in channel.connection_hops) is registered in the context.

Reply:
I also wonder why they do it.. cause the verification checks the transactions, and if present all written data is assumed correct .. but I let it in case I miss something.

Reply:
For the local connection id it makes sense. But here we construct the counterparty connection hops that include the counterparty ids (`ccid` is the only one).